core/bin/zksync_api/src/fee_ticker/validator/watcher.rs

```diff
@@ -2,74 +2,94 @@ use std::collections::HashMap;
 use std::sync::Arc;
 use std::time::Instant;
 
-use bigdecimal::{BigDecimal, Zero};
-use serde::{Deserialize, Serialize};
+use super::types as CGTypes;
+use bigdecimal::{BigDecimal, FromPrimitive};
+use serde::de::DeserializeOwned;
 use tokio::sync::Mutex;
 use zksync_types::{Address, Token};
 
-use crate::fee_ticker::ticker_api::REQUEST_TIMEOUT;
+use crate::fee_ticker::{ticker_api::REQUEST_TIMEOUT, validator::types::Contract};
 
 #[async_trait::async_trait]
 pub trait TokenWatcher {
     async fn get_token_market_volume(&mut self, token: &Token) -> anyhow::Result<BigDecimal>;
 }
 
-/// Watcher for Uniswap protocol
-/// https://thegraph.com/explorer/subgraph/uniswap/uniswap-v2
 #[derive(Clone)]
-pub struct UniswapTokenWatcher {
+pub struct CoinGeckoTokenWatcher {
     client: reqwest::Client,
     addr: String,
+    chain_id: u8,
     cache: Arc<Mutex<HashMap<Address, BigDecimal>>>,
 }
 
-impl UniswapTokenWatcher {
-    pub fn new(addr: String) -> Self {
+impl CoinGeckoTokenWatcher {
+    pub fn new(addr: String, chain_id: u8) -> Self {
         Self {
             client: reqwest::Client::new(),
             addr,
+            chain_id,
             cache: Default::default(),
         }
     }
-    async fn get_market_volume(&mut self, address: Address) -> anyhow::Result<BigDecimal> {
-        // Uniswap has graphql API, using full graphql client for one query is overkill for current task
-        // TODO https://linear.app/matterlabs/issue/ZKS-413/support-full-version-of-graphql-for-tokenvalidator
-        let start = Instant::now();
-
-        let query = format!("{{token(id: \"{:#x}\"){{untrackedVolumeUSD}}}}", address);
 
+    async fn get<T>(&self, query: &str) -> Result<T, anyhow::Error>
```

Comment:
```suggestion
    async fn get_from_client<T>(&self, query: &str) -> Result<T, anyhow::Error>
```

Reply:
I don't think this adds readability to get(query) tbh